C# reference: Consolidated jump statements (#27410)
dotnet/docs: docs/csharp/language-reference/statements/jump-statements.md

---
title: "Jump statements - C# reference"
description: "Learn about C# jump statements: break, continue, return, and goto."
ms.date: 12/06/2021
f1_keywords:
  - "break_CSharpKeyword"
  - "continue_CSharpKeyword"
  - "return_CSharpKeyword"
  - "goto_CSharpKeyword"
helpviewer_keywords:
  - "break statement [C#]"
  - "break keyword [C#]"
  - "continue statement [C#]"
  - "continue keyword [C#]"
  - "return statement [C#]"
  - "return keyword [C#]"
  - "goto statement [C#]"
  - "goto keyword [C#]"
---
# Jump statements (C# reference)

The following statements unconditionally transfer control:

- The [`break` statement](#the-break-statement): terminates the closest enclosing [iteration statement](iteration-statements.md) or [`switch` statement](selection-statements.md#the-switch-statement).
- The [`continue` statement](#the-continue-statement): starts a new iteration of the closest enclosing [iteration statement](iteration-statements.md).

> - The [`return` statement](#the-return-statement): terminates execution of the function in which it appears and returns control to the caller.

Author: Is there any actual feedback from readers about this possible confusion? The current reference article doesn't have any such feedback. Spec, which is usually very specific, doesn't mention that topic. I think exiting method clearly implies exiting a loop. I suggest to keep it simple.

> - The [`return` statement](#the-return-statement): terminates execution of the function in which it appears and returns control to the caller.

Author: Also, the ref return example has the `return` statement inside a loop. I guess that may be enough?

> - The [`goto` statement](#the-goto-statement): transfers control to a statement that is marked by a label.

Author: The [spec about jump statements](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#jump-statements) also includes the `throw` statement in the list of jump statements. Should we cover `throw` in this article (and then to have the separate article only for `try` and its variants)?

Reviewer: I would expect to see `throw` here, as it in my mind is a jump statement. But I defer to @BillWagner

Reviewer 2: I thought about this for a while. There are good arguments on both sides.  In the end, I think it will be more clear to link to an article about `throw`, `try`, `catch` and `finally` (and probably exception filters) from this article than to cover `throw` here.

My reasoning is that `throw` is strongly related to `try`, `catch`, and `finally`. 

The `throw` statement is a jump statement in the spec. But what about the `throw` expression? It's not even a *statement*. If I were looking for `throw`, I'd expect to find information on both the *statement* and the *expression*. (Although, I think they will stay separate in the spec because of the 1st level headings)

Author: @BillWagner thank you; so, I've only added the link to the article about `throw` for now. (Later, I'll consolidate all exception-related statements into one article.)

> ## The `break` statement

Reviewer 2: There are a few other cases here that could be explored. This may be another case where not qualifying a statement can be clarifying. I'm not sure yet.

- What about nested `switch` statements? A `break` only exits the inner-most switch. 
- What about a `break` inside a loop in the statements for a *switch-condition*. The `break` will exit the loop, but not the `switch` statement.
- Any level of nesting of those conditions, among others.

I don't have alternative, and it may not be necessary, but it's worth thinking through. From customer questions, the nested loops example *is* important.

Author: From the current feedback to the `break` article, I've concluded that nested loops and `switch` inside a loop are worth mentioning. As for other variants, I assume (1) they are less frequent (and maybe should be refactored), (2) may be reasoned about by analogy from the existing examples. So, I don't think the article should explore other cases. What do you think @BillWagner?

:::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/BreakStatement.cs" id="BasicExample":::

In nested loops, the `break` statement terminates only the innermost loop that contains it, as the following example shows:

:::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/BreakStatement.cs" id="NestedLoop":::

When you use the `switch` statement inside a loop, a `break` statement at the end of a switch section transfers control only out of the `switch` statement. The loop that contains the `switch` statement is unaffected, as the following example shows:

:::code language="csharp" source="snippets/jump-statements/BreakStatement.cs" id="SwitchInsideLoop":::

## The `continue` statement

The `continue` statement starts a new iteration of the closest enclosing [iteration statement](iteration-statements.md) (that is, `for`, `foreach`, `while`, or `do` loop), as the following example shows:

:::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/ContinueStatement.cs" id="BasicExample":::

## The `return` statement

> The `return` statement terminates execution of the function in which it appears and returns control and the function's result, if any, to the caller. The `return` statement can be any of the following two forms:

Reviewer 2: Regarding the next question, I might consider something like this:

```suggestion
The `return` statement terminates execution of the function in which it appears and returns control and the function's result, if any, to the caller. The `return` statement has an expression following the `return` keyword when the member produce a value. The `return` statement must not have an expression when the member doesn't prodcue a value. When the `return` statement has an expression, that expression must be convertible to the type returned by the member. The only exception to this rule is `async` functions, where the expression must be convertible to the type argument for <xref:System.Threading.Tasks.Task%601?displayProperty=nameWithType>. Return statements in `async` methods that return <xref:System.Threading.Tasks.Task?displayProperty=nameWithType> must not have an expression.
```

Author: @BillWagner thank you, I've rewritten this section based on your suggestion. Please review it again.

- Without expression, as in the following example:

  :::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/ReturnStatement.cs" id="WithoutExpression":::

  You can use the `return` statement without expression in a function that doesn't compute a value, that is:

  - a [method](../../programming-guide/classes-and-structs/methods.md), [local function](../../programming-guide/classes-and-structs/local-functions.md), or [lambda expression](../operators/lambda-expressions.md) with the result type [`void`](../builtin-types/void.md)
  - an [async](../keywords/async.md) method, local function, or lambda expression with the result type <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.ValueTask>
  - the `set` accessor of a [property](../../programming-guide/classes-and-structs/properties.md) or an [indexer](../../programming-guide/indexers/index.md)
  - a [constructor](../../programming-guide/classes-and-structs/constructors.md)
  - a [finalizer](../../programming-guide/classes-and-structs/finalizers.md)

>   - the `add` and `remove` accessors of an [event](../../programming-guide/events/index.md)

Author: This list is very spec-like. Alternatives:
(1) Remove it at all, leave only the introduction sentence
(2) Shorten it, make it not exhaustive, that is, mention only method and properties as examples
(3) Do nothing, leave as it is
(4) Something else

@BillWagner what do you think?

Reviewer 2: I think it will be easier to read using the introductory sentence structure. Consider something like my suggestion on line 51.

  You typically use the `return` statement without expression to terminate a function early. If a function doesn't contain the `return` statement, it terminates after its last statement is executed.

- With expression, as in the following example:

  :::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/ReturnStatement.cs" id="WithExpression":::

  You use the `return` statement with expression in a function that computes a value, that is:

>   - a [method](../../programming-guide/classes-and-structs/methods.md), [local function](../../programming-guide/classes-and-structs/local-functions.md), or [lambda expression](../operators/lambda-expressions.md) with a non-void result type
>   - an [async](../keywords/async.md) method, local function, or lambda expression with the result type <xref:System.Threading.Tasks.Task%601> or <xref:System.Threading.Tasks.ValueTask%601>
>   - the `get` accessor of a [property](../../programming-guide/classes-and-structs/properties.md) or an [indexer](../../programming-guide/indexers/index.md)
>   - an [overloaded operator](../operators/operator-overloading.md)

Reviewer: ```suggestion
  - A [method](../../programming-guide/classes-and-structs/methods.md), [local function](../../programming-guide/classes-and-structs/local-functions.md), or [lambda expression](../operators/lambda-expressions.md) with a non-void result type.
  - An [async](../keywords/async.md) method, local function, or lambda expression with the result type <xref:System.Threading.Tasks.Task%601> or <xref:System.Threading.Tasks.ValueTask%601>.
  - The `get` accessor of a [property](../../programming-guide/classes-and-structs/properties.md) or an [indexer](../../programming-guide/indexers/index.md).
  - An [overloaded operator](../operators/operator-overloading.md).
```

I think we'd want sentences for the list here.

>   By default, the `return` statement returns the value of an expression. Beginning with C# 7.0, you can return a reference to a variable. To do that, use the `return` statement with the [`ref` keyword](../keywords/ref.md), as the following example shows:

Reviewer 2: Is it worth mentioning that `ref return` is not valid in an `async` method?

Author: I don't think so, as I've added the link to the article about ref returns just after the example.

>   :::code language="csharp" source="snippets/jump-statements/ReturnStatement.cs" id="RefReturn":::

Reviewer: ```suggestion
  :::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/ReturnStatement.cs" id="RefReturn":::
```

I verified that this one does work in the interactive experience.

  For more information about ref returns, see [Ref returns and ref locals](../../programming-guide/classes-and-structs/ref-returns.md).

> ## The `goto` statement

Reviewer 2: I think it's worth mentioning that there are restrictions on the target label of a `goto` statement.

The [standard](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/statements.md#13104-the-goto-statement) has some very elegant language to define the restriction:

> If a label with the given name does not exist in the current function member, or if the `goto` statement is not within the scope of the label, a compile-time error occurs.

This rule means you can't use `goto` to transfer control *into* a `try` block, `catch` block, `finally` block, or other nested scope.

It might be sufficient to say you can't use `goto` to transfer control out of the current member, or into any nested scope, such as a `try` block.

Author: @BillWagner thanks, I've updated the section accordingly.

The `goto` statement transfers control to a statement that is marked by a label, as the following example shows:

:::code language="csharp" interactive="try-dotnet-method" source="snippets/jump-statements/GotoStatement.cs" id="NestedLoops":::

As the preceding example shows, you can use the `goto` statement to get out of a nested loop.

> [!TIP]

> > When you work with nested loops, consider refactoring separate loops into separate methods. That may lead to a simpler, more readable code without the `goto` statement.

Reviewer: Indeed, a `goto` is often a code smell.
